File: Matlab/hospital_mean.m
%hospital_mean.m

function [men_table, women_table] = hospital_mean(patients,files_patients,folder_path, ...
     hospital_id,  men_table, women_table)


hmr_patient_id_mapping = containers.Map(...
    {'hmr_001_landmark', 'hmr_002_landmark', 'hmr_003_landmark', 'hmr_004_landmark', ...
     'hmr_005_landmark', 'hmr_006_landmark', 'hmr_007_landmark', 'hmr_008_landmark', ...
     'hmr_019_landmark', 'hmr_027_landmark', 'HN-HRM-041'}, ...
    {'HN-HMR-001', 'HN-HMR-002', 'HN-HMR-003', 'HN-HMR-004', ...
     'HN-HMR-005', 'HN-HMR-006', 'HN-HMR-007', 'HN-HMR-008', ...
     'HN-HMR-019', 'HN-HMR-027','HN-HMR-041'});

chum_patient_id_mapping = containers.Map(...
    {'landmark_distances_mm_1', 'landmark_distances_mm_3', 'landmark_distances_mm_4', ...
    'landmark_distances_mm_5','landmark_distances_mm_6','landmark_distances_mm_7' ...
    'landmark_distances_mm_8','landmark_distances_mm_11','landmark_distances_mm_13','landmark_distances_mm_14', ...
    'HN-CHUM-061','HN-CHUM-063', 'HN-CHUM-064'}, ...
    {'HN-CHUM-001', 'HN-CHUM-003', 'HN-CHUM-004',... 
     'HN-CHUM-005', 'HN-CHUM-006','HN-CHUM-007','HN-CHUM-008','HN-CHUM-011', ...
     'HN-CHUM-013','HN-CHUM-014','HN-CHUM-061', ...
     'HN-CHUM-063', 'HN-CHUM-064'} );

chus_patient_id_mapping = containers.Map( ...
    {'HNCHUS001_distances', 'HNCHUS002_distances', 'HNCHUS003_distances', ...
    'HNCHUS004_distances', 'HNCHUS005_distances'}, ...
    {'HN-CHUS-001', 'HN-CHUS-002', 'HN-CHUS-003', 'HN-CHUS-004', 'HN-CHUS-005'});
hgj_patient_id_mapping = containers.Map(...
    {'HN-HGJ-005', 'HN-HGJ-006', 'HN-HGJ-008', ...
    'HN-HGJ-009','HN-HGJ-012','HN-HGJ-091'}, ...
    {'HN-HGJ-005', 'HN-HGJ-006', 'HN-HGJ-008',... 
     'HN-HGJ-009', 'HN-HGJ-012', 'HN-HGJ-091' } );


% Read CSV file


    for i = 1:length(files_patients)
                file_path = fullfile(folder_path, files_patients(i).name);
                patient_landmark = readtable(file_path);
                name_id = strsplit(files_patients(i).name, '.');
                name_id = name_id{1};

                if (hospital_id == 1) 
                      patient_id = chum_patient_id_mapping(name_id);
                      
                end
                if (hospital_id == 2)
                    patient_id = hmr_patient_id_mapping(name_id);

                end
               
                if (hospital_id == 3) 
                      patient_id = hgj_patient_id_mapping(name_id);
                end
                if (hospital_id == 4) 
                      patient_id = chus_patient_id_mapping(name_id);
                end
                patient = patients(strcmp(patients{:, 'Patient #'}, patient_id), :);
               
                
                % Determine age category
                if (patient.Age < 25)
                    age = 'Under 25';
                    
                elseif (patient.Age >= 25 && patient.Age < 60)
                    age = '25-60';
                elseif (patient.Age >= 60)
                    age = 'Over 60';
                end
                hospital_patient = table(string(patient_id), patient.Sex, string(age), patient_landmark.Distance_mm', 'VariableNames', {'PatientID', 'Sex', 'AgeCategory', 'Soft Tissue Thickness'});

                
                % Update mean and standard deviation for men
                if (strcmp(patient.Sex, 'M'))
                   
                    men_table = [men_table; hospital_patient];

            
                end
            
                % Update mean and standard deviation for women
                if (strcmp(patient.Sex, 'F'))
            
                    
                    women_table = [women_table; hospital_patient];
                    
                end
    end
end

File: Soft tissue thickness final files/Python/hospital_mean.py
import pandas as pd
import numpy as np
import os

def hospital_mean(patients, files_patients, folder_path, hospital_id, men_table, women_table):
    hmr_patient_id_mapping = {
        'hmr_001_landmark': 'HN-HMR-001', 'hmr_002_landmark': 'HN-HMR-002', 'hmr_003_landmark': 'HN-HMR-003', 'hmr_004_landmark': 'HN-HMR-004',
        'hmr_005_landmark': 'HN-HMR-005', 'hmr_006_landmark': 'HN-HMR-006', 'hmr_007_landmark': 'HN-HMR-007', 'hmr_008_landmark': 'HN-HMR-008',
        'hmr_019_landmark': 'HN-HMR-019', 'hmr_027_landmark': 'HN-HMR-027', 'HN-HRM-041': 'HN-HMR-041'
    }

    chum_patient_id_mapping = {
        'landmark_distances_mm_1': 'HN-CHUM-001', 'landmark_distances_mm_3': 'HN-CHUM-003', 'landmark_distances_mm_4': 'HN-CHUM-004',
        'landmark_distances_mm_5': 'HN-CHUM-005', 'landmark_distances_mm_6': 'HN-CHUM-006', 'landmark_distances_mm_7': 'HN-CHUM-007',
        'landmark_distances_mm_8': 'HN-CHUM-008', 'landmark_distances_mm_11': 'HN-CHUM-011', 'landmark_distances_mm_13': 'HN-CHUM-013',
        'landmark_distances_mm_14': 'HN-CHUM-014', 'HN-CHUM-061': 'HN-CHUM-061', 'HN-CHUM-063': 'HN-CHUM-063', 'HN-CHUM-064': 'HN-CHUM-064'
    }

    chus_patient_id_mapping = {
        'HNCHUS001_distances': 'HN-CHUS-001', 'HNCHUS002_distances': 'HN-CHUS-002', 'HNCHUS003_distances': 'HN-CHUS-003',
        'HNCHUS004_distances': 'HN-CHUS-004', 'HNCHUS005_distances': 'HN-CHUS-005', 'HNCHUS006_distances': 'HN-CHUS-006',
        'HNCHUS007_distances': 'HN-CHUS-007', 'HNCHUS008_distances': 'HN-CHUS-008', 'HNCHUS009_distances': 'HN-CHUS-009',
        'HNCHUS0010_distances': 'HN-CHUS-010', 'HNCHUS0011_distances': 'HN-CHUS-011', 'HNCHUS0012_distances': 'HN-CHUS-012', 'HNCHUS0013_distances': 'HN-CHUS-013'
    }

    hgj_patient_id_mapping = {
        'HN-HGJ-005': 'HN-HGJ-005', 'HN-HGJ-006': 'HN-HGJ-006', 'HN-HGJ-008': 'HN-HGJ-008',
        'HN-HGJ-009': 'HN-HGJ-009', 'HN-HGJ-012': 'HN-HGJ-012', 'HN-HGJ-091': 'HN-HGJ-091'
    }

    patient_id_mapping = {
        0: chum_patient_id_mapping,
        1: hmr_patient_id_mapping,
        2: hgj_patient_id_mapping,
        3: chus_patient_id_mapping
    }

    current_mapping = patient_id_mapping[hospital_id]

    for file_path in files_patients:
        if isinstance(file_path, str):
            patient_landmark = pd.read_csv(file_path)
            name_id = os.path.basename(file_path).split('.')[0]

            if name_id in current_mapping:
                patient_id = current_mapping[name_id]
            else:
                continue

            # Stampa i nomi delle colonne per la verifica
            print(f"Columns in {file_path}: {patient_landmark.columns}")

            if 'Distance_mm' not in patient_landmark.columns:
                print(f"Warning: 'Distance_mm' not found in {file_path}. Skipping this file.")
                continue

            patient = patients.loc[patients['Patient #'] == patient_id]

            if patient.empty:
                continue

            if patient['Age'].values[0] < 25:
                age = 'Under 25'
            elif 25 <= patient['Age'].values[0] < 60:
                age = '25-60'
            else:
                age = 'Over 60'

            hospital_patient = pd.DataFrame({
                'PatientID': [patient_id],
                'Sex': [patient['Sex'].values[0]],
                'AgeCategory': [age],
                'Soft Tissue Thickness': [patient_landmark['Distance_mm'].values]
            })

            if patient['Sex'].values[0] == 'M':
                men_table = pd.concat([men_table, hospital_patient], ignore_index=True)
            else:
                women_table = pd.concat([women_table, hospital_patient], ignore_index=True)

    return men_table, women_table
